### ISSUE TITLE
Error Message: Joplin Resources Directory does not exist

### ISSUE THREAD

**Issue by pxcask** [d4vydm/CherryTree2Joplin#4]
Hey there, happy new year :) And thanks for the promising tool!

This is what the skript responds ... :

`me@laptop:~/path/to/pyskript$ python3 ct2joplin.py /path/to/exported/directory/
Traceback (most recent call last):
  File "ct2joplin.py", line 480, in <module>
    main(ct_html_dir)
  File "ct2joplin.py", line 335, in main
    resource_id_name = get_joplin_resourceid_filename_mapping(ct_html_dir)
  File "ct2joplin.py", line 80, in get_joplin_resourceid_filename_mapping
    cur.execute("select id,title from resources")
sqlite3.OperationalError: no such table: resources
`
Hmmm, what's wrong? With the -ji option it says
`Exiting... The joplin resources directory does not exist.`

New directories are created as well as one .html file in a new directory called htmlexport.
Joplin CLI and Joplin are installed from the snap store on Ubuntu 20.04 

Cheers from Germany
pxcask

**Comment by d4vydm** [comment]
Hi,

Do you have the Joplin sqlite db in ~/.config/joplin-desktop/database.sqlite ?
Or might be that the file is located elsewhere due to snap install.

I was thinking maybe the DB is not initialised, which might happen when you open Joplin for the first time. That would then be a requirement for the script to function correctly.

Can you check/try both?

Thanks!

**Comment by pxcask** [comment]
Hi, the file is in ~/snap/joplin-desktop/current/.config/joplin-desktop as it is a snap install.
I guess I can just mount -bind the folder it into the ~/.config/ ... folder, right?
Thanks!

**Comment by d4vydm** [comment]
Yes should be fine.

I'll see if I can add an option to enter the link to the joplin db.

**Comment by d4vydm** [comment]
@pxcask I added the option to pass the config dir of your joplin.

Use:
`-jc ~/snap/joplin-desktop/current/.config/joplin-desktop`

And you get:
```
Joplin sqlite file: /home/xxxx/snap/joplin-desktop/current/.config/joplin-desktop/database.sqlite
Joplin resource dir:/home/xxxx/snap/joplin-desktop/current/.config/joplin-desktop/resources
...
```

Should work

**Comment by pxcask** [comment]
If I use 
`python3 ct2joplin.py -jc ~/snap/joplin-desktop/current/.config/joplin-desktop /path/to/exported/directory/`
it says
`usage: ct2joplin.py [-h] [--version] [-v] [-kp] [-ku] [-ji] directory
ct2joplin.py: error: unrecognized arguments: -jc /path/to/exported/directory/`

**Comment by d4vydm** [comment]
@pxcask Did you do a fetch of the new version of the script?

**Comment by pxcask** [comment]
Yes, the version from 20:40:41 on 08Jan (today)

**Comment by d4vydm** [comment]
Whats the output of `python3 ct2joplin.py -h`

**Comment by pxcask** [comment]
Hi d4vydm, this is the output of '`python3 ct2joplin.py -h'`

`usage: ct2joplin.py [-h] [--version] [-v] [-kp] [-ku] [-ji] [-jc JOPLIN_CONFIG_DIR] directory

Import CherryTree notes into Joplin. Dependencies: pandoc, joplin cli. (Author: D4vyDM)

positional arguments:
  directory             directory containing the exported CherryTree html files

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -v, --verbose
  -kp, --keep_parent_notes
                        By default the program assumes only leaf nodes have notes. With this flag, also import parent nodes note data.
  -ku, --keep_underscores
                        By default underscores in node names are replaced with spaces. With this flag we keep underscores in node names.
  -ji, --joplin_import  Import converted data directly into Joplin
  -jc JOPLIN_CONFIG_DIR, --joplin_config_dir JOPLIN_CONFIG_DIR
                        Pass the path of the Joplin configuration directory

CherryTree exports parent-child-leaf relationship using double dash (e.g. Parent--Child1--Child2--Leaf.html). Spaces are replaced with underscore (Parent--Child1--Child2--
Leaf_with_spaces.html). Special chars are not escaped (Parent--Child1--Child_with_(paranthesis)_and_stuff--Leaf.html). Joplin has the concept of notebooks and notes. A notebook itself
has no content, only notes and other notebooks. Cherrytree does allow each node to have content. By default we assume parent nodes do not have content, as Joplin sees these as
notebooks. Only leafs have content and are considered notes. If you do wish to keep the content of the parent nodes, use the -kp flag The program converts the flat directory HTML export
from CherryTree to a structured MD files tree We use the joplin import function to import from a directory`

**Comment by netweaver1970** [comment]
Hi, I'm having the same error message 

![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/ccf758fc-40b1-40d5-8d82-645cc4d7895a)

but my db file is -I presume- in the correct location. Strange though I have 2 ?? joplin was installed from app store and joplin_cli was installed via brew. 

![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/2a2b4922-ef3f-41d4-8daa-537dbcf322bf)

corrected.
![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/538ee337-8efd-4ae0-ab8a-e6a43a35b8b0)

I downloaded joplin, joplin-cli and your python script today.

Any idea?

Thanks.

**Comment by netweaver1970** [comment]
I got a bit further, I saw that the CLI was clearly accessing it's own profile, different from the Desktop app profile. So I created a symlink from the Desktop directory to the CLI directory and then I ran into the next issue. But with the same -non descriptive- text when executing the py script.

When running the joplin command by itself, I got more info. Seems to be most likely related to the db version mismatch between Desktop and CLI. See my request here on the Joplin Discourse forum [https://discourse.joplinapp.org/t/db-version-mismatch-with-latest-available-joplin-downloads/32167](https://discourse.joplinapp.org/t/db-version-mismatch-with-latest-available-joplin-downloads/32167)

@d4vydm maybe TRY CATCH the messages coming back from every DB invocation via the CLI and report back to the enduser, in case not as expected. That would help troubleshooting the -otherwise vague- issues.

Thanks.

**Comment by d4vydm** [comment]
@netweaver1970 Thanks for the feedback. Been a while I worked on this script. 

The error you're seeing happens after the temporary import of the files to to get the reference ID's Joplin uses for files and images.

Maybe this is happening because your joplin DB hasn't been initialized yet, which could be due to the difference between Joplin-cli and joplin-desktop db locations.

I could input a check that the Joplin DB has been initialized first before starting.

The symlink fixed the issue and the import was ok?

**Comment by netweaver1970** [comment]
Not yet, the desktop version creates a db version 43 while the cli can only handle up to db version 42. So waiting for an updated cli version.

**Comment by netweaver1970** [comment]
Laurent was very helpful, he upgraded joplin-cli to version 2.12.1, which uses the same DB version as the desktop.
Yesterday also the homebrew devs were very fast in packaging up the new version up in homebrew, I installed it fine and the joplin-cli now also uses the same DB as the desktop. all good so far.
Now back to square 1, the conversion. Update later.

**Comment by netweaver1970** [comment]
> @netweaver1970 Thanks for the feedback. Been a while I worked on this script.
> 
> The error you're seeing happens after the temporary import of the files to to get the reference ID's Joplin uses for files and images.
> 
> Maybe this is happening because your joplin DB hasn't been initialized yet, which could be due to the difference between Joplin-cli and joplin-desktop db locations.
> 
> I could input a check that the Joplin DB has been initialized first before starting.
> 
> The symlink fixed the issue and the import was ok?

no luck yet, this is the reponse:
```
xx@Geerts-MBP Downloads % python3 ct2joplin.py -v export_HTML -ji -jc ~/.config/joplin-desktop/database.sqlite
Exiting... MacOSX platform not implemented.
Joplin sqlite file:
Joplin resource dir:
Exiting... The joplin resources directory does not exist.
```

Joplin-cli starts perfectly fine, using the desktop DB directory (via directory symlink).
the script isn't giving much info, even in the verbose mode?

**Comment by netweaver1970** [comment]
and without the -ji flags:
```
...
Coping file export_HTML/images/45-3.png to export_HTML/htmlimport/images
Coping file export_HTML/images/115-1.png to export_HTML/htmlimport/images
Coping file export_HTML/images/45-2.png to export_HTML/htmlimport/images
Coping file export_HTML/images/26-1.png to export_HTML/htmlimport/images
Coping file export_HTML/images/104-8.png to export_HTML/htmlimport/images
Creating import dir: export_HTML/htmlimport/EmbeddedFiles
Coping file export_HTML/EmbeddedFiles/29-pgconduitthreads.zip to export_HTML/htmlimport/EmbeddedFiles
Importing notebook: export_HTML/htmlimport
Importing notes...

...ok
Traceback (most recent call last):
  File "/Users/xxx/Downloads/ct2joplin.py", line 515, in <module>
    main(ct_html_dir)
  File "/Users/xxx/Downloads/ct2joplin.py", line 357, in main
    resource_id_name = get_joplin_resourceid_filename_mapping(ct_html_dir)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xxx/Downloads/ct2joplin.py", line 92, in get_joplin_resourceid_filename_mapping
    cur.execute("select id,title from resources")
sqlite3.OperationalError: no such table: resources
```

**Comment by netweaver1970** [comment]
actually something DID happen (I did try a few variants of the command line, hence the multiple imports)
Is this how it's supposed to look? No tree structure anymore after imports?
can a default more toned-down template be used (as most of my notes are/were very low coloring/fidelity? I used CT more as a scrapbook of references, values and pictures.
What about the resources sqlite3 error at the end of the import?

![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/e88bfdd3-bf0f-46ce-9f44-94de31b34ca3)

**Comment by netweaver1970** [comment]
some PyCharm debug later... I'm not a python expert but I can read/debug and write a little bit, so please give me some slack on this. 
First, it seems, line 92-93, variable JOPLIN_SQLLITE_FILE is empty, so it's normal that this error happens!
![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/92319758-a2a3-42c8-abb6-e0285eb1742c)

It's being filled here, of course in my case it will be empty -by design- , since MacOS (darwin) is not supported :D
![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/1ad5ef47-66d2-4514-8e61-0942d0d8677d)


When just copying the same code as for unix in the 'darwin' elif check, all starts working. Also, when just adding it to the unix elif check and just deleting the now-defunct lines 459-462, that's easiest:
![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/b4210501-daff-4fa1-9d23-2432de16dc9c)

So, basically ALWAYS include the -ji parameter, and when specifying the config directory parameter, use the full path, I had issues with using ~ in the path. But that might have been my issue.

Now things look like this:
![image](https://github.com/d4vydm/CherryTree2Joplin/assets/18028618/7eef1e35-2742-41f8-94ef-ffb1ce5543c1)

@d4vydm Is it meant to show this number behind every note after conversion or is that still a bug?
@pxcask In case you're still interested, you'll have to debug yourself as well to see what could be the issue but, since you're on Ubuntu, I suspect you simply didn't add the -ji parameter, so the necessary resource variable didn't get filled. 
[edit]Or you used tilde in your path, it seems the libs used don't like that. It happened to me as well.
